frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/BaseResetBranchToRemoteAction.java

```diff
@@ -0,0 +1,185 @@
+        LOG.debug(() -> "Resetting '${branchName}' branch");
+        try (AccessToken ignored = DvcsUtil.workingTreeChangeStarted(project, TASK_TITLE)) {
+          GitLineHandler resetHandler = new GitLineHandler(myProject, gitRepository.getRoot(), GitCommand.RESET);
+          resetHandler.addParameters("--keep");
+
+          Option<BaseGitMacheteBranch> branchOption = macheteRepository.getBranchByName(branchName);
+          assert branchOption.isDefined() : "Can't get branch '${branchName}' from Git Machete repository";
+          Option<IGitMacheteRemoteBranch> remoteTrackingBranchOption = branchOption.get().getRemoteTrackingBranch();
+          if (remoteTrackingBranchOption.isDefined()) {
+            resetHandler.addParameters(remoteTrackingBranchOption.get().getPointedCommit().getHash());
+          } else {
+            String message = "Branch '${branchName}' doesn't have remote tracking branch, so cannot be reset";
+            LOG.warn(message);
+            VcsNotifier.getInstance(project).notifyWarning(VCS_NOTIFIER_TITLE, message);
+            return;
+          }
+
+          resetHandler.endOptions();
+
+          // Check if branch to reset is not current branch - if isn't then checkout
+          Option<String> currentBranchOption = getCurrentBranchNameIfManaged(anActionEvent);
+          if (currentBranchOption.isEmpty() || !currentBranchOption.get().equals(branchName)) {
+            LOG.debug(() -> "Checkout to branch '${branchName}' is needed");
+            // Checking out given branch
+            CheckoutSelectedBranchAction.doCheckout(branchName, gitRepository, project, indicator);
+
+            // Check again if we are in branch to reset to be sure that checkout was successful
+            // This time we are using git4idea because GitMacheteRepository is immutable and it would return previous branch
+            @Nullable GitLocalBranch localBranch = gitRepository.getCurrentBranch();
+            if (localBranch == null || !localBranch.getName().equals(branchName)) {
+              LOG.error("Checkout to branch ${branchName} failed");
+              VcsNotifier.getInstance(project).notifyError(VCS_NOTIFIER_TITLE,
+                  "Error occurred during checkout of the branch that was to be reset! Operation aborted.");
+              return;
+            } else {
+              LOG.debug(() -> "Checkout to branch '${branchName}' successful");
+            }
+          }
+
+          GitCommandResult result = Git.getInstance().runCommand(resetHandler);
+          if (!result.success()) {
+            LOG.error(result.getErrorOutputAsJoinedString());
+            VcsNotifier.getInstance(project).notifyError(VCS_NOTIFIER_TITLE, result.getErrorOutputAsHtmlString());
+            return;
+          }
+        }
+
+        // If we are here this means that all went good
+        VcsNotifier.getInstance(project).notifySuccess("Branch '${branchName}' reset to remote");
```

Comment:
Hmmm when testing locally I've noticed that GitMacheteGraphTable isn't refreshed (the `(diverged from origin)` doesn't disappear). Maybe call refresh action here then?

Reply:
Strange, it's worked for me...

Reply:
Yeah, might be undeterministic :(

Reply:
Restore previous refreshing methods invocation